README.md
# CampusPedia Internship Time Tracker

Simple tim tracking apps that helps campuspedia to track their intern working hours on campuspedia.

## Technologies used

1. **Backend:** Laravel and Firebase
2. **Frontend:** Bootstrap

## Contributing

*I'am sorry, currently contribution are closed due this project is my assignmen towards Campuspedia Internship Progran. But, in the meantime, I will make this project feels like OSS :smile:

## References and Readbooks

1. https://www.mynotepaper.com/laravel-realtime-crud-using-google-firebase/
2. https://medium.com/@arifulislam_ron/a-quick-guide-to-implement-firebase-web-push-notification-in-laravel-ac11ba3260b4
3. https://medium.com/@josiahoyahaya/building-a-crud-app-with-google-firebase-using-phpfirebase-laravel-vuejs-larafirebase-2d6bc1732784
4. https://dzone.com/articles/the-ultimate-vuejs-amp-laravel-crud-tutorial
5. https://www.mynotepaper.com/integrate-google-firebase-in-laravel
6. https://laravel.com/docs/7.x/homestead
7. https://medium.com/@dani.ardians/laravel-cara-menghubungkan-laravel-5-7-dengan-firebase-c305b6920579
8. https://stackoverflow.com/questions/15592058/convert-complex-excel-formula-to-javascript
9. https://stackoverflow.com/questions/42730666/real-time-javascript-calculation-based-on-form-inputs
10. https://stackoverflow.com/questions/10804042/calculate-time-difference-with-javascript
11. https://stackoverflow.com/questions/4467539/javascript-modulo-gives-a-negative-result-for-negative-numbers
12. https://stackoverflow.com/questions/13091599/time-tracking-javascript-inc-jquery-or-php
13. https://stackoverflow.com/questions/30430182/jquery-looping-through-checkbox-and-storing-in-an-array
14. https://firebase.google.com/docs/
15. https://laravel.com/docs/7.x
16. https://api.jquery.com/
17. https://learn.jquery.com/
18. 
19. 
20. 

## LICENSE

This project licensed under MIT License

CHANGELOG
## [ Unrelease ]
- Applying design styles from Carbon DS

## [0.1.1-alpha] - 202-05-19
### Added
- Admin dashboard for viewing interns job reports
- Add Detailed fields in job reporting
- Connection to Firebase Realtime Database

### Fixed
- Firebase realtime databse key push

### Known Bugs
- Hardcoded breaktime timetracked

## [0.1.0-alpha] - 2020-05-18
### Added
- Simple time track that show how much time spent by the users from html time input

config/services.php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Third Party Services
    |--------------------------------------------------------------------------
    |
    | This file is for storing the credentials for third party services such
    | as Mailgun, Postmark, AWS and more. This file provides the de facto
    | location for this type of information, allowing packages to have
    | a conventional file to locate the various service credentials.
    |
    */

    'mailgun' => [
        'domain' => env('MAILGUN_DOMAIN'),
        'secret' => env('MAILGUN_SECRET'),
        'endpoint' => env('MAILGUN_ENDPOINT', 'api.mailgun.net'),
    ],

    'postmark' => [
        'token' => env('POSTMARK_TOKEN'),
    ],

    'ses' => [
        'key' => env('AWS_ACCESS_KEY_ID'),
        'secret' => env('AWS_SECRET_ACCESS_KEY'),
        'region' => env('AWS_DEFAULT_REGION', 'us-east-1'),
    ],

    'firebase' => [
        'apiKey' => 'AIzaSyBcgTYnvkUI4lSFgx3XyY7NtrKmr8lu_7U',
        'authDomain' => 'campuspedia-timetrack.firebaseapp.com',
        'databaseURL' => 'https://campuspedia-timetrack.firebaseio.com',
        'projectId' => 'campuspedia-timetrack',
        'storageBucket' => 'campuspedia-timetrack.appspot.com',
        'messagingSenderId' => '240849287434',
        'appId' => '1:240849287434:web:f8af6dc74305403d91a2f3'
    ]

];


resources/views/admin.blade.php
<!doctype html>
<html lang="en">
<head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    <!-- Bootstrap CSS-->
    <link defer rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css">
     
    <title>Daily Reports App</title>

</head>
<body>

<div class="container" style="margin-top: 50px;">

<h4 class="text-center">Daily Job Reporting Admin Dashboard</h4><br>

    <h5>Reports</h5>
    <table class="table table-bordered">
        <tr>
            <th>Name</th>
            <th>Date Submit</th>
            <th width="180" class="text-center">Action</th>
        </tr>
        <tbody id="tbody">

        </tbody>
    </table>
</div>

<!-- Update Model -->
<form action="" method="POST" class="users-update-record-model form-horizontal">
    <div id="update-modal" data-backdrop="static" data-keyboard="false" class="modal fade" tabindex="-1" role="dialog" aria-labelledby="custom-width-modalLabel"
         aria-hidden="true">
        <div class="modal-dialog modal-dialog-centered" style="width:200%;">
            <div class="modal-content" style="overflow: hidden;">
                <div class="modal-header">
                    <h4 class="modal-title" id="custom-width-modalLabel">Informasi Pekerjaan</h4>
                    <button type="button" class="close" data-dismiss="modal"
                            aria-hidden="true">×
                    </button>
                </div>
                <div class="modal-body" id="updateBody">

                </div>
                <div class="modal-footer">
                    <button type="button" class="btn btn-light"
                            data-dismiss="modal">Close
                    </button>
                </div>
            </div>
        </div>
    </div>
</form>


{{--Firebase Tasks--}}
<script src="https://code.jquery.com/jquery-3.4.0.min.js"></script>
<script src="https://www.gstatic.com/firebasejs/5.10.1/firebase.js"></script>
<script>
    // Initialize Firebase
    var config = {
        apiKey: "{{ config('services.firebase.apiKey') }}",
        authDomain: "{{ config('services.firebase.authDomain') }}",
        databaseURL: "{{ config('services.firebase.databaseURL') }}",
        storageBucket: "{{ config('services.firebase.storageBucket') }}",
    };
    firebase.initializeApp(config);

    var database = firebase.database();

    var lastIndex = 0;

    // Get Data
    firebase.database().ref('reporting/').on('value', function (snapshot) {
        var value = snapshot.val();
        var htmls = [];
        $.each(value, function (index, value) {
            if (value) {
                htmls.push('<tr>\
        		<td>' + value.name + '</td>\
        		<td>' + value.timestamp + '</td>\
        		<td><button data-toggle="modal" data-target="#update-modal" class="btn btn-info updateData" data-id="' + index + '">View data</button>\
        	</tr>');
            }
            lastIndex = index;
        });
        $('#tbody').html(htmls);
        $("#submitUser").removeClass('disabled');
    });

    // Update Data
    var updateID = 0;
    $('body').on('click', '.updateData', function () {
        updateID = $(this).attr('data-id');
        firebase.database().ref('reporting/' + updateID).once('value', function (snapshot) {
            var values = snapshot.val();
            console.log(values)
            var updateData = '<div class="form-group">\
		        <label for="first_name" class="col-md-12 col-form-label font-weight-bold">Name</label>\
		        <div class="col-md-12">\
		            <input id="first_name" type="text" class="form-control-plaintext" name="name" value="'+ values.name +'">\
		        </div>\
            </div>\
            <div class="form-group">\
		        <label for="first_name" class="col-md-12 col-form-label font-weight-bold">Nomor Internship</label>\
		        <div class="col-md-12">\
		            <input id="first_name" type="text" class="form-control-plaintext" name="name" value="'+ values.uid + '">\
		        </div>\
            </div>\
            <div class="form-group">\
		        <label for="first_name" class="col-md-12 col-form-label font-weight-bold">Pekerjaan</label>\
		        <div class="col-md-12">\
		            <input id="first_name" type="text" class="form-control-plaintext" name="name" value="'+ values.today_jobs +'">\
		        </div>\
            </div>\
            <div class="form-group">\
		        <label for="first_name" class="col-md-12 col-form-label font-weight-bold">Deskripsi</label>\
		        <div class="col-md-12">\
		            <input id="first_name" type="text" class="form-control-plaintext" name="name" value="'+ values.today_jobs_problems +'">\
		        </div>\
            </div>\
            <div class="form-group">\
		        <label for="first_name" class="col-md-12 col-form-label font-weight-bold">Waktu Bekerja </label>\
		        <div class="col-md-12">\
		            <input id="first_name" type="text" class="form-control-plaintext" name="name" value="'+ values.workingtime_start + ' until ' + values.workingtime_end +'">\
		        </div>\
            </div>\
            <div class="form-group">\
		        <label for="first_name" class="col-md-12 col-form-label font-weight-bold">Lama Jam Kerja (bersih)</label>\
		        <div class="col-md-12">\
		            <input id="first_name" type="text" class="form-control-plaintext" name="name" value="'+ values.workhour_net +'">\
		        </div>\
            </div>\
            <div class="form-group">\
		        <label for="first_name" class="col-md-12 col-form-label font-weight-bold">Tanggal laporan dibuat</label>\
		        <div class="col-md-12">\
		            <input id="first_name" type="text" class="form-control-plaintext" name="name" value="'+ values.timestamp +'">\
		        </div>\
            </div>';

            $('#updateBody').html(updateData);
        });
    });

</script>

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js"></script>

</body>
</html>

resources/views/submitreport.blade.php
<!doctype html>
<html lang="en">

<head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css">

    <title>Daily Reports App</title>

</head>

<body>

    <div class="container" style="margin-top: 50px;">

        <h4 class="text-center">Daily Job Reporting</h4><br>

        <h5>Add new reports</h5>
        <div class="card card-default">
            <div class="card-body">
                <form id="addReport" class="users-update-record-model form-horizontal" method="POST" action="">
                    <div class="form-group">
                        <label for="intern_name" class="col-md-12 col-form-label">Name</label>
                        <input id="intern_name" type="text" class="form-control" name="name" placeholder="Name" required
                            autofocus>
                    </div>

                    <div class="form-group">
                        <label for="uid" class="col-md-12 col-form-label">Nomor Internship</label>
                        <input id="uid" type="text" class="form-control" name="uid" placeholder="Nomor Internship kamu" required
                            autofocus>
                    </div>

                    <div class="form-group">
                        <label for="today_jobs" class="col-md-12 col-form-label">Apa yang kamu kerjakan hari ini?</label>
                        <input id="today_jobs" type="text" class="form-control" name="today jobs"
                            placeholder="Apa aja nih....">
                    </div>

                    <div class="form-group">
                        <label for="today_jobs_problems" class="col-md-12 col-form-label">Deskripsi hari ini</label>
                        <textarea id="today_jobs_problems" type="text" name="today_jobs_desc" class="form-control"
                            form="addCustomer" value="Masalah hari ini">Coba ceritain kendala hari ini dong...</textarea>
                    </div>

                    <div class="form-group">
                        <label for="workingtime_start" class="col-md-12 col-form-label">Hari ini, kamu mulai kerja jam
                            berapa?</label>
                        <input id="workingtime_start" type="time" class="form-control" name="workstart"
                            placeholder="HH:MM">
                    </div>

                    <div class="form-group">
                        <label for="workingtime_end" class="col-md-12 col-form-label">Selesainya jam berapa?</label>
                        <input id="workingtime_end" type="time" class="form-control" name="workend"
                            placeholder="HH:MM">
                    </div>

                    <div class="form-group">
                        <label class="col-md-12 col-form-label ">Kamu Pastinya istirahat dong? Jam berapa kalo boleh
                            tau</label>

                        <input type="checkbox" id="breaktime-noon" class="breaktimes" value="Siang">
                        <label for="breaktime-noon" class="">Break Siang (12:00 - 13:00)</label>
                        <br>
                        <input type="checkbox" id="breaktime-afternoon" class="breaktimes" value="Sore">
                        <label for="breaktime-afternoon" class="">Break Sore (15:00 - 16:00)</label>
                        <br>
                        <input type="checkbox" id="breaktime-night" class="breaktimes" value="Malam">
                        <label for="breaktime-night" class="">Break Malam (18:00 - 19:00)</label>

                    </div>

                    <button id="sendReport" type="button" class="btn btn-primary mb-2">Add New Reports</button>
                </form>
            </div>
        </div>

    {{--Firebase Tasks--}}
    <script src="https://code.jquery.com/jquery-3.4.0.min.js"></script>
    <script src="https://www.gstatic.com/firebasejs/5.10.1/firebase.js"></script>
    <script>
        // Initialize Firebase
        var config = {
            apiKey: "{{ config('services.firebase.apiKey') }}",
            authDomain: "{{ config('services.firebase.authDomain') }}",
            databaseURL: "{{ config('services.firebase.databaseURL') }}",
            storageBucket: "{{ config('services.firebase.storageBucket') }}",
        };
        firebase.initializeApp(config);
        var database = firebase.database();

        let breaktime = 0,
            nettHour = 0,
            totalWork = 0

        // Submit Report Function
        $('#sendReport').on('click', function () {

            // Get values from the forms
            start = $("#workingtime_start").val()
            end = $("#workingtime_end").val()


            // Start Calculation for the work hours and breaktimes
            breakTimeCalculator(start, end)
            workTime_total(start, end)
            workTime_nett(totalWork, breaktime)

            /* if (breaktime != breaktime_formcheck) {
                alert("Waktu istirahat dan pilihan istirahatmu tidak cocok!")
                return;
            } */

            // Data Object to be pushed into firebase
            reportDetails = {
                name: $("#intern_name").val(),
                uid : $("#uid").val(),
                today_jobs : $("#today_jobs").val(),
                today_jobs_problems : $("#today_jobs_problems").val(),
                workingtime_start : $("#workingtime_start").val(),
                workingtime_end : $("#workingtime_end").val(),
                breaktimes : breaktime,
                workhour_total : totalWork,
                workhour_net : nettHour,
                timestamp: Date(),
            }


            // Push data to the firebase
            firebasePush(reportDetails)

            // Prints data to the consoles for debugging
            console.log("Mulai kerja jam: " + start)
            console.log("Berhenti kerja jam: " + end)
            console.log("Breaktime selama: " + breaktime)
            console.log("==================")
            console.log("Total Worktime: " + totalWork)
            console.log("Net Hour: " + nettHour)
            console.log(reportDetails)

            // Cleans Input Forms
            $("#addReport input").val("");
        });

        function breakTimeCalculator(startTime, endTime) {
            // Kerja sebelum jam 12AM sampai diatas jam 7PM
            if (startTime < "12:00" && endTime > "19:00") {
                breaktime = 3
            }
            // Kerja sebelum jam 12AM sampai diatas jam 5PM
            else if (startTime < "12:00" && endTime > "17:00") {
                breaktime = 2
            }
            // Kerja sebelum jam 12AM sampai diatas jam 1PM
            else if (startTime < "12:00" && endTime > "13:00") {
                breaktime = 1
            }
            // Kerja sesudah jam 1PM & jam 4PM sampai diatas jam 7 PM
            else if ((startTime > "13:00" && startTime > "16:00") && endTime > "19:00") {
                breaktime = 2
            }
            // Kerja sesudah jam 1PM & jam 4PM sampai diatas jam 5 PM
            else if ((startTime > "13:00" && startTime > "16:00") && endTime > "17:00") {
                breaktime = 2
            }
            // Kerja sesudah jam 5PM & jam 6PM sampai diatas jam 7 PM
            else if ((startTime > "17:00" && startTime > "18:00") && endTime > "19:00") {
                breaktime = 2
            }
            return breaktime;
        }

        function workTime_nett(totalWork, breaktime) {
            nettHour = totalWork - breaktime

            return nettHour;
        }

        function timeStringToFloat(time) {
            var hoursMinutes = time.split(/[.:]/);
            var hours = parseInt(hoursMinutes[0], 10);
            var minutes = hoursMinutes[1] ? parseInt(hoursMinutes[1], 10) : 0;
            return hours + minutes / 60;
        }

        function workTime_total(startTime, endTime) {
            timeWork_start = timeStringToFloat(startTime)
            timework_end = timeStringToFloat(endTime)
            totalWork = timework_end - timeWork_start

            return totalWork;
        }

        function firebasePush(reports) {
            firebase.database().ref('reporting/').push(reports,
            function (error) {
                if (error) {
                    alert("Could not send the report! Please try again")
                } else {
                    alert("Report Submitted. Thank you")
                }
            });
        }

    </script>

    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js"></script>

</body>

</html>
